include/aws/common/linked_list.h

```diff
@@ -0,0 +1,91 @@
+#ifndef AWS_LINKED_LIST_H
+#define AWS_LINKED_LIST_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <stdint.h>
+
+struct aws_linked_list_node {
+    struct aws_linked_list_node *next;
+    struct aws_linked_list_node *prev;
+};
+
+struct aws_linked_list {
```

Comment:
Consider having the 'list' itself be a node, and having the list be circular. This is how it's done in the linux kernel and leads to a dramatic simplification in the linked list code - a lot of the branches can be removed. It also means we can be a lot more flexible in how we manage spliced-out sections of the list (something that isn't really handled with this API), as any single node can define a list on its own.